File: geekbang/algorithm-models/5.3-快速排序代码模板/quickSort.go
package main

// Golang

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	return nil
}

func sortArray(nums []int) []int {
	quickSort(nums, 0, len(nums)-1)
	return nums
}

func quickSort(arr []int, l, r int) {
	if l >= r {
		return
	}
	pivot := partition(arr, l, r)
	quickSort(arr, l, pivot)
	quickSort(arr, pivot+1, r)
}

func partition(a []int, l, r int) int {
	pivot := l + rand.Intn(r-l+1)
	pivotVal := a[pivot]

	for l <= r {
		for a[l] < pivotVal {
			l++
		}
		for a[r] > pivotVal {
			r--
		}
		if l == r {
			break
		}
		if l < r {
			a[l], a[r] = a[r], a[l]
			l++
			r--
		}
	}
	return r
}

File: geekbang/algorithm-models/5.3-快速排序代码模板/quickSort.cpp
// C/C++


class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        quickSort(nums, 0, nums.size() - 1);
        return nums;
    }


    void quickSort(vector<int>& arr, int l, int r) {
        if (l >= r) return;
        int pivot = partition(arr, l, r);
        quickSort(arr, l, pivot);
        quickSort(arr, pivot + 1, r);
    }


    int partition(vector<int>& a, int l, int r) {
        int pivot = l + rand() % (r - l + 1);
        int pivotVal = a[pivot];


        while (l <= r) {
            while (a[l] < pivotVal) l++;
            while (a[r] > pivotVal) r--;
            if (l == r) break;
            if (l < r) {
                int temp = a[l]; a[l] = a[r]; a[r] = temp;
                l++; r--;
            }
        }


        return r;
    }
};
